### DISCUSSION TITLE
Dijkstra

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#33] [Esercizi A-L]
Dato un grafo $G = (V, E)$, data la funzione $\omega : E \to \mathbb{R}^+$ che associa ad ogni arco $e \in E$ un peso, definiamo la distanza fra due vertici $x, y \in V$ come il cammino con il peso totale minimo.

Scrivere lo pseudocodice di un algoritmo che trova la distanza fra due vertici $x, y$ nel grafo pesato, in $O(n(n + m))$ o in $O((n + m)log(n))$

**Comment by interestingmaneuver** [comment]
Bonus challenge perché per l'originale basta semplicemente presentare l'algoritmo.

Cosa succede se $\omega$ è definita invece come $\omega : E \to \mathbb{R}$?

**Comment by Frisbon** [comment]
### Algoritmo di Dijkstra Ottimizzato
```python
def Dijkstra(G,x,y):
    D = [inf,...,inf]; D[x] = 0
    H = heap()
    for nodo in V(G):
        H.insert( (nodo, D[nodo]) ) #tupla 
    while H is not empty:
        x = H.pop_min()
        D[x] = x.value #prende il valore dalla tupla
        for y in x.ad:
            stima = D[x] + peso(x,y)
            if (y in H) and D[y] > stima:
                H.update_value(x, stima) 
    return D[y]
```

> @interestingmaneuver 
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/29de174d-951e-49a9-b17e-1b594bf8b63c)

I pesi possono assumere sia valori negativi che lo zero
_(potrei sbagliare)_
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/ece6c7c3-6891-49f5-8549-acdcef6331d6)

**Reply by interestingmaneuver** [reply]
Cosa succede se

$$c(a, d) = 5$$

$$c(b, d) = 10$$

$$c(c, d) = -3$$

*Bonus*: se invece il costo tra b e d rimane 1, che costo ha il cammino minimo tra a e c?

**Comment by daromacarlo** [comment]
```c
//ALGORITMO CHE OPERA IN O((n+m)log(n))

def Opt_Dijkstra(G, x, y)
	dist = [inf,...,inf]
	dist[x] = 0
	haep = [value, key]
	heap.add(x, 0)
	while queue is not empty{
		v = min(heap.key)
		for w in v.adiacenti{
		if dist[w] > dist[v] + d(v,w){
			dist[w] = dist[v] + d(v,w)
			w.update_key(dist[v] + d(v,w))
			heap.add(w, dist[w])
			}
		}
		heap.remove(v)
	}
	return dist[y]
``` 
```c
//ALGORITMO CHE OPERA IN O((n+m)n)

def unOpt_Dijkstra(G,x)
	dist = [0,...,0]
	dist[x] = 0
	R += [x]
	while R != V(G){
		min = infinito
		min_arco = null
		for (x,y) | x in R and y not in R{
			if dist[x] + d(x,y) < min{
				min = dist[x] + d(x,y)
				min_arco = (x,y)
			}
		}
		R.add(min_arco[1])
		dist[min_arco[1]] = min
	}
	return dist[y]
```
